### ISSUE TITLE
Interesting Bug About Variable Indexes

### ISSUE THREAD

**Issue by Blueyescat** [bensku/Skript#913]
## Test Skript:
```vb
set {test::%type of player's tool%} to "test"
set {test::%player's tool%} to "test"
loop {test::*}:
	send "%loop-index% - %loop-value%"
```
## Result:
![image](https://user-images.githubusercontent.com/19474570/31311743-761da24c-abbb-11e7-91be-3c5e72f2347d.png)
![image](https://user-images.githubusercontent.com/19474570/31311780-276926de-abbc-11e7-9d10-5f36cf359164.png)

## Versions
**Skript**: 2.2-dev31c
**Server**: PaperSpigot 1.12 (b1169)` and `1.12.2 (b1234)

### Note:
Some checks like ``{test::%type of player's tool%} is set`` works if i'm holding the same items.

**Comment by Pikachu920** [comment]
I think this is intended behavior, but I'm not sure. I actually like it -- it happens with locations too (when used in variables the world is also printed) which is nice for parsing.

**Comment by TheBentoBox** [comment]
Yeah, don't think this is a bug, it's how objects are represented when stored as an index. It's useful because it lets you parse out the type from the index.

**Comment by Blueyescat** [comment]
What? Are you serious? ..

**Comment by TheBentoBox** [comment]
Yes? What's the problem?

**Comment by bensku** [comment]
Seems to be unintentional change in some recent version.

**Comment by Pikachu920** [comment]
So, i'm looking at this and I totally cannot understand how this could ever have been an unintentional change. the toVariableString was purposefully changed at some point (this has been like this since at least 2013 https://github.com/bensku/Skript/blob/ae20a83b91768523f69dc410b03586188e4dd360/src/main/java/ch/njol/skript/classes/data/BukkitClasses.java#L239). I think that it should just stay at this point as this issue has been around for quite a while and it provides utility in recreating a Location and an ItemStack/Type. What do you guys think?

**Comment by Blueyescat** [comment]
It doesn't even work fine for ``player's tool``. It shouldn't be a feature, even in Skript, it is ugly. Variable indexes always known as texts. ``{%tool%}`` and ``"%tool%"`` works same. Already there is no way to get the object from the variable index because it is a text, still you need to use parse. Also it works only for some specified Skript types.

**Comment by Pikachu920** [comment]
{%expression%} and "%expression%" aren't always meant to work the same way. thats why a type's parser has seperate methods for both. The reason this is more useful for parsing is because the world is included, which it usually isnt (e.g. when broadcasted). The reason it doesn't work for `player's tool` is because tool (annoyingly) returns a Slot, not a ItemStack/ItemType.

**Comment by Blueyescat** [comment]
They used to work same until this weird issue. If you want world, use ``"%world%: %location%"``. This is not useful or something, I have not seen it used as technically in Skript. It looks like unnecessary.

**Comment by bensku** [comment]
Next build has rather big changes to slot handling. I advise you to check again when it hits.

**Comment by TheLimeGlass** [comment]
Can this be closed now?

**Comment by Blueyescat** [comment]
Maybe when that unnecessary thing is removed

**Comment by TheLimeGlass** [comment]
Use type of player's tool

**Comment by Blueyescat** [comment]
No lol

**Comment by moraedu** [comment]
While most do not agree these should be human-readable, from this issue we had a rather lengthy and contrived discussion in the discord, from which I personally concluded these should be taken into account for the `toVariableString` of these specific cases:
* for `ItemType` when there's a backing `ItemStack` such as the case for `player's tool` or any slot. Be the same as `ItemStack`'s `toVariableString`.
* for `ItemType` when it is an alias, it should remain as it is right now.
* for `ItemStack` we should consider removing durability from the `toVariableString` (doubt this will happen as it is a breaking change).
* for `Slot` when it is `EquipmentSlot.OFFHAND` or `EquipmentSlot.MAINHAND`, it should just be `ItemStack`'s `toVariableString`.
* for other types of `Slot` it should match `ItemStack`? What about the slot indices?

Also unrelated but also somewhat on topic:
* `Location`'s `toVariableString` should be parseable at the `ParseContext.COMMAND` (soon to be `ParseContext.PARSE`)

**Comment by Fusezion** [comment]
> * for `Slot` when it is `EquipmentSlot.OFFHAND` or `EquipmentSlot.MAINHAND`, it should just be `ItemStack`'s `toVariableString`.
> * for other types of `Slot` it should match `ItemStack`? What about the slot indices?

All slots are saved as ItemStacks in variables, the whole usage of slot toVariableString needs to be updated. not just tool or offhand.
